src/background/providers/cloudflare.ts

```diff
@@ -227,59 +254,71 @@ export class CloudflareProvider implements Provider {
     handleBeforeRequest(
         details: chrome.webRequest.WebRequestBodyDetails,
     ): chrome.webRequest.BlockingResponse | void {
-        const url = new URL(details.url);
-
+        // Only issue tokens for POST requests that contain 'application/x-www-form-urlencoded' data.
         if (
             details.requestBody === null ||
-            details.requestBody === undefined ||
-            details.requestBody.formData === undefined
+            details.requestBody === undefined
         ) {
             return;
         }
 
-        const hasQueryParams = QUALIFIED_QUERY_PARAMS.some((param) => {
-            return url.searchParams.has(param);
-        });
-        const hasBodyParams = QUALIFIED_BODY_PARAMS.some((param) => {
-            return details.requestBody !== null && param in details.requestBody.formData!;
-        });
-        if (!hasQueryParams || !hasBodyParams) {
+        const url = new URL(details.url);
+        const formData: { [key: string]: string[] | string } = details.requestBody.formData || {};
+
+        // Only issue tokens on the issuing website.
+        if (!isIssuingHostname(ALL_ISSUING_CRITERIA.HOSTNAMES, url)) {
+            return;
+        }
+
+        // Only issue tokens when the pathname passes defined criteria.
+        if (!isQualifiedPathname(ALL_ISSUING_CRITERIA.PATHNAMES, url)) {
+            return;
+        }
+
+        // Only issue tokens when querystring parameters pass defined criteria.
+        if (!areQualifiedQueryParams(ALL_ISSUING_CRITERIA.QUERY_PARAMS, url)) {
+            return;
+        }
+
+        // Only issue tokens when POST data parameters pass defined criteria.
+        if (!areQualifiedBodyFormParams(ALL_ISSUING_CRITERIA.BODY_PARAMS, formData)) {
             return;
         }
 
         const flattenFormData: { [key: string]: string[] | string } = {};
-        for (const key in details.requestBody.formData) {
-            if (details.requestBody.formData[key].length == 1) {
-                const [value] = details.requestBody.formData[key];
+        for (const key in formData) {
+            if (Array.isArray(formData[key]) && (formData[key].length === 1)) {
+                const [value] = formData[key];
                 flattenFormData[key] = value;
             } else {
-                flattenFormData[key] = details.requestBody.formData[key];
+                flattenFormData[key] = formData[key];
             }
         }
 
-        (async () => {
-            // Issue tokens.
-            const tokens = await this.issue(details.url, flattenFormData);
-            // Store tokens.
-            const cached = this.getStoredTokens();
-            this.setStoredTokens(cached.concat(tokens));
+        // delay the request to issue tokens until next tick of the event loop
+        setTimeout(
```

Comment:
Why do you change to `setTimeout` over here?

Reply:
because (if memory serves) that's a blocking API method.. and there's no need for the browser to wait for the secondary request to finish; using a next tick allows the triggered code to run asynchronously.

maybe there's a good reason to not do it this way, but it struck me as a better idea while I was touching that portion of code.

Reply:
actually.. my answer just now was wrong..

```javascript
(async () => {...})()
```

wouldn't block..

I think my reason for adding the timeout in the first place had to do with hCaptcha. I don't entirely understand how their process works.. but it seems (from trial and error) very clear that the intercepted request needs to complete __before__ the triggered request to issue tokens can be made.. otherwise it fails. So.. i don't know if removing this timeout would break hCaptcha or not.. but there's a good chance that it might.

Reply:
Of course.. these comments about hCaptcha don't apply to the plugin for the Cloudflare provider; the update to Cloudflare was just because I wanted the 2x plugins to keep as nearly identical as possible.. for symmetry.. so any differences in logic can be easily grok'd by diff'ing the two.

Reply:
__update:__
Regarding my earlier comment/assertion that hCaptcha requires the timeout.. and also..   
Regarding my earlier comments about testing the ES5-transpiled release of the extension in older browsers..  
I just ran a very interesting test that addresses both topics:
* added a [polyfill library](https://cdnjs.cloudflare.com/ajax/libs/core-js/3.20.3/minified.min.js) to the _public_ directory and loaded it from both the background page and the popup
* using Chrome 30:
  - packaged the extension to a crx2 and installed
    * no errors or warnings.. everything looks good
  - solved an _hCaptcha_ and inspected network traffic
    * request from normal window tab to solve captcha: 200
    * request from background page to issue tokens: 403
      - `{"success":false,"error-codes":["invalid-data"]}`
      - which is the response that occurs when the request to issue tokens happens _before_ the POST to solve
* made a quick change to the timeout in the background page: increased from `0` to `2500` milliseconds
  - repacked the crx2 and reinstalled
  - solved another _hCaptcha_ and... earned 5x tokens
    * it works in Chrome 30! ..how cool is that?
* the Cloudflare provider doesn't work out-of-the-box
  - haven't looked into it yet
  - guessing it's because the TLS certificate isn't trusted by this browser
    * which would make the necessary `window.crypto` API unavailable
    * I'll need to add its root authority.. and retest..

Reply:
I just added another commit to this branch.. it's tagged as `v3.4.0` with a corresponding crx under [releases](https://github.com/warren-bank/fork-crx-PrivacyPass-challenge-bypass/releases)

* adds hook functions that delegate to providers for:
  - [`chrome.webRequest.onCompleted`](https://developer.chrome.com/docs/extensions/reference/webRequest/#event-onCompleted)
  - [`chrome.webRequest.onErrorOccurred`](https://developer.chrome.com/docs/extensions/reference/webRequest/#event-onErrorOccurred)
* reverts the Cloudflare provider to remove the previously added timeout
* refactors the hCaptcha provider to move the code that sends issuing requests
  - from a timeout in 'handleBeforeRequest'
    * ie: the hook for [`chrome.webRequest.onBeforeRequest`](https://developer.chrome.com/docs/extensions/reference/webRequest/#event-onBeforeRequest)
  - to 'handleOnCompleted' and 'handleOnErrorOccurred':
    * ie: the new hooks for [`chrome.webRequest.onCompleted`](https://developer.chrome.com/docs/extensions/reference/webRequest/#event-onCompleted) and [`chrome.webRequest.onErrorOccurred`](https://developer.chrome.com/docs/extensions/reference/webRequest/#event-onErrorOccurred), respectively
* updates tests to adjust for revisions
  - all pass

this design is:
1. much cleaner
2. much more reliable, since the timing of the secondary request depends upon completion of the 1st

the result is:
* stable and fully functional

Reply:
PS: this new release was built using the improved build scripts, which automates the insertion of the ES6+ polyfill library (that I mentioned previously) into the CRX2 extension.. so
* the CRX3 does _not_ include polyfill
  - since it only runs in newer browsers
* the CRX2 does include the polyfill
  - since it only runs in legacy browsers
  - this backports support for some crazy old browsers..
    * the primary bottleneck now is whether or not the website operated by each provider will also support the older browsers